File: services/characterService.js
const axios = require('axios');
const cheerio = require('cheerio');

const BASE_URL = 'https://fr.finalfantasyxiv.com';
const LODESTONE_URL = `${BASE_URL}/lodestone/character`;

// Structure des Data Centers et leurs serveurs
const dataCenters = {
  'Chaos': [
    'Cerberus', 'Louisoix', 'Moogle', 'Omega', 
    'Phantom', 'Ragnarok', 'Sagittarius', 'Spriggan'
  ],
  'Light': [
    'Alpha', 'Lich', 'Odin', 'Phoenix', 
    'Raiden', 'Shiva', 'Twintania', 'Zodiark'
  ]
};

/**
 * Recherche des personnages en fonction du nom et, optionnellement, du serveur ou du Data Center
 * @param {string} name - Nom du personnage
 * @param {string} server - Serveur spécifique (optionnel)
 * @param {string} dataCenter - Data Center (optionnel)
 * @returns {Promise<Array>} - Liste des personnages trouvés
 */
exports.searchCharacters = async (name, server = '', dataCenter = '') => {

  try {
    // Construire l'URL de recherche
    let searchUrl = `${LODESTONE_URL}/?q=${encodeURIComponent(name)}`;
    
    // Si un serveur spécifique est fourni, l'ajouter à l'URL de recherche
    if (server) {
      searchUrl += `&worldname=${encodeURIComponent(server)}`;
    }
    
    // Si un Data Center spécifique est fourni, l'ajouter à l'URL de recherche
    if (dataCenter) {
      searchUrl += `&dataCenter=${encodeURIComponent(dataCenter)}`;
    }
    // Effectuer la requête
    const response = await axios.get(searchUrl);
    const $ = cheerio.load(response.data);

    // Récupérer la liste des personnages
    let characters = [];
    
    $('.entry__link').each((i, element) => {
      const $element = $(element);
      const $characterName = $element.find('.entry__name');
      const $characterServer = $element.find('.entry__world');
      const $characterImage = $element.find('.entry__chara__face img');
      
      // Extraire l'ID depuis l'URL
      const characterUrl = $element.attr('href');
      const characterId = characterUrl.split('/').filter(Boolean).pop();
      const characterServerName = $characterServer.text().trim();

      const [server, dataCenter] = characterServerName.split(" [").map(part => part.replace("]", "").trim());
      
      characters.push({
        id: characterId,
        name: $characterName.text().trim(),
        server: server,
        avatar: $characterImage.attr('src'),
        profileUrl: `${BASE_URL}${characterUrl}`,
        dataCenter: dataCenter
      });
    });
    
    // Si un Data Center est fourni, filtrer les résultats par Data Center
    if (dataCenter && !server) {
      characters = characters.filter(character => 
        character.dataCenter === dataCenter || 
        serverBelongsToDataCenter(character.server, dataCenter)
      );
    }
    
    return characters;
  } catch (error) {
    console.error('Erreur lors du scraping des personnages:', error);
    throw new Error(`Échec du scraping: ${error.message}`);
  }
};

/**
 * Détermine à quel Data Center appartient un serveur
 * @param {string} serverName - Nom du serveur
 * @returns {string} - Nom du Data Center, ou chaîne vide si non trouvé
 */
function getDataCenterByServer(serverName) {
  for (const [dataCenter, servers] of Object.entries(dataCenters)) {
    if (servers.includes(serverName)) {
      return dataCenter;
    }
  }
  return '';
}

/**
 * Vérifie si un serveur appartient à un Data Center
 * @param {string} serverName - Nom du serveur
 * @param {string} dataCenterName - Nom du Data Center
 * @returns {boolean} - true si le serveur appartient au Data Center
 */
function serverBelongsToDataCenter(serverName, dataCenterName) {
  return dataCenters[dataCenterName] && dataCenters[dataCenterName].includes(serverName);
}

exports.getCharacterDetails = async (id) => {
  try {
    // Construire l'URL du personnage
    const characterUrl = `${LODESTONE_URL}/${id}/`;
    
    // Effectuer la requête
    const response = await axios.get(characterUrl);
    const $ = cheerio.load(response.data);
    
    // Extraire les informations de base
    const name = $('.frame__chara__name').text().trim();
    const avatar = $('.frame__chara__face img').attr('src');
    const title = $('.frame__chara__title').text().trim();
    const portrait = $('.character__detail__image img').attr('src');
    
    // Extraire le serveur et le data center
    const serverInfo = $('.frame__chara__world').text().trim();
    const [server, dataCenter] = serverInfo.split(' [');
    
    // Extraire les informations de classe et niveau
    const jobs = [];
    
    $('.character__level__list li').each((i, element) => {
      const $element = $(element);
      const jobName = $element.find('img').attr('data-tooltip');
      const jobLevel = $element.text().trim()
      
      if (jobName && jobLevel) {
        jobs.push({
          name: jobName,
          level: parseInt(jobLevel, 10) || 0
        });
      }


    });
    
    // Extraire les informations d'entreprise (FC)
    let freeCompany = null;
    const fcElement = $('.character__freecompany__name a');
    
    if (fcElement.length > 0) {
      const fcName = fcElement.text().trim();
      const fcUrl = fcElement.attr('href');
      const fcId = fcUrl.split('/').filter(Boolean).pop();
      
      freeCompany = {
        id: fcId,
        name: fcName,
        url: `${BASE_URL}${fcUrl}`
      };
    }
    
    // Rassembler toutes les informations
    const character = {
      id,
      name,
      title,
      server: server,
      dataCenter: dataCenter ? dataCenter.replace(']', '') : '',
      avatar,
      portrait,
      profileUrl: characterUrl,
      jobs,
      freeCompany
    };
    
    return character;
  } catch (error) {
    console.error('Erreur lors du scraping des détails du personnage:', error);
    throw new Error(`Échec du scraping: ${error.message}`);
  }
};

File: services/character_service.py
import asyncio
import re
from aiohttp import ClientSession, ClientResponseError
from bs4 import BeautifulSoup, Tag
from config.data_centers import DataCenters
from flask import g
from http import HTTPStatus
from requests import get, HTTPError
from urllib.parse import quote


def search_characters_service(name: str, server: str = "") -> list[dict]:
    """
    Search for characters based on name and optionally server

    :param name: character's name
    :param server: server (optional)
    :return: Found characters
    """
    try:
        # Loading datacenters
        DataCenters.load_data()

        # Check if server is valid
        if server and server not in DataCenters.get_all_servers():
            raise ValueError(f"Server '{server}' doesn't exist. Check the name")

        # Building URL
        search_url = f"{g.base_url}/lodestone/character/?q={quote(name)}&worldname={quote(server)}"

        response = get(search_url)
        response.raise_for_status()
        soup = BeautifulSoup(response.text, "html.parser")

        characters = []
        # Getting characters entries
        for entry in soup.select(".entry__link"):
            name_elem = entry.select_one(".entry__name")
            server_elem = entry.select_one(".entry__world")
            image_elem = entry.select_one(".entry__chara__face img")
            lang_elem = entry.select_one(".entry__chara__lang")

            character_url = str(entry.get("href"))
            # Extracting ID from URL
            parts = [part for part in character_url.split("/") if part]
            character_id = parts[-1] if parts else None

            # Extracting information from the server and data center
            character_server_text = (
                server_elem.get_text(strip=True) if server_elem else ""
            )
            server_data = character_server_text.split(" [")
            server_name = server_data[0].strip() if len(server_data) > 0 else ""
            data_center_value = (
                server_data[1].replace("]", "").strip() if len(server_data) > 1 else ""
            )

            characters.append(
                {
                    "id": character_id,
                    "name": name_elem.get_text(strip=True) if name_elem else "",
                    "server": server_name,
                    "lang": lang_elem.get_text(strip=True) if lang_elem else "",
                    "avatar": image_elem.get("src") if image_elem else "",
                    "profileUrl": f"{g.base_url}{character_url}",
                    "dataCenter": data_center_value,
                }
            )
        return characters

    except ValueError as error:
        raise error

    except Exception as error:
        print("Error while scraping characters: ", error)
        raise Exception(f"Scraping failed: {str(error)}")


def get_character_details_service(character_id: int) -> dict:
    """
    Retrieves a character's details by ID.

    :param character_id: character's ID
    :return: Dictionary containing character details
    """
    try:
        # Building URL
        character_url = f"{g.base_url}/lodestone/character/{character_id}/"
        response = get(character_url)
        response.raise_for_status()
        soup = BeautifulSoup(response.text, "html.parser")

        # Extracting basic information
        name_elem = soup.select_one(".frame__chara__name")
        avatar_elem = soup.select_one(".frame__chara__face img")
        title_elem = soup.select_one(".frame__chara__title")
        portrait_elem = soup.select_one(".character__detail__image img")

        # Extraction of the server and the data center
        server_info_elem = soup.select_one(".frame__chara__world")
        server_info = server_info_elem.get_text(strip=True) if server_info_elem else ""
        server_parts = server_info.split(" [")
        server = server_parts[0].strip() if len(server_parts) > 0 else ""
        data_center_value = (
            server_parts[1].replace("]", "").strip() if len(server_parts) > 1 else ""
        )

        # Extracting class and level information
        jobs = []
        for li in soup.select(".character__level__list li"):
            img_elem = li.find("img")

            if isinstance(img_elem, Tag):
                job_name = img_elem.get("data-tooltip", "") if img_elem else ""
                job_img = img_elem.get("src") if img_elem else ""

            # Clean up the class name
            if isinstance(job_name, str) and re.search(r"[\/()]", job_name):
                job_name = re.split(r"[\/()]", job_name)[0].strip()

            job_level_text = li.get_text(strip=True)

            try:
                job_level = int(job_level_text)
            except ValueError:
                job_level = 0
            if job_name and job_level_text:
                jobs.append({"name": job_name, "level": job_level, "image": job_img})

        # Extracting free company information (FC)
        free_company = None
        fc_elem = soup.select_one(".character__freecompany__name a")
        if fc_elem:
            fc_name = fc_elem.get_text(strip=True)
            fc_url = str(fc_elem.get("href"))
            fc_parts = [part for part in fc_url.split("/") if part]
            fc_id = fc_parts[-1] if fc_parts else None
            free_company = {
                "id": fc_id,
                "name": fc_name,
                "url": f"{g.base_url}{fc_url}",
            }

        # Gather all the information
        character = {
            "id": character_id,
            "name": name_elem.get_text(strip=True) if name_elem else "",
            "title": title_elem.get_text(strip=True) if title_elem else "",
            "server": server,
            "dataCenter": data_center_value,
            "avatar": avatar_elem.get("src") if avatar_elem else "",
            "portrait": portrait_elem.get("src") if portrait_elem else "",
            "profileUrl": character_url,
            "freeCompany": free_company,
            "jobs": jobs,
        }

        return character

    except HTTPError as error:
        raise error

    except Exception as error:
        print("Error while scraping characters: ", error)
        raise Exception(f"Scraping failed: {str(error)}")


async def get_character_achievements_service(character_id: int) -> dict:
    """
    Search character achievements by ID

    :param character_id: character's ID
    :return: Dictionary containing character details
    """
    try:
        # Building URL
        character_achievements_url = (
            f"{g.base_url}/lodestone/character/{character_id}/achievement"
        )

        async with ClientSession() as session:
            response = await fetch(session, character_achievements_url)

            if response:
                soup = BeautifulSoup(response, "html.parser")

            achievements = {}

            # Récupération des catégories principales
            category_elements = soup.select("nav.achievement__category dt")[1:]

            # Lancement de toutes les requêtes pour récupérer les sous-catégories en parallèle
            tasks = []
            for entry_category_achievement in category_elements:
                category_elem = entry_category_achievement.select_one("a")
                if category_elem:
                    category_name = category_elem.get_text(strip=True)
                    category_urlArray = str(category_elem["href"]).split("/")
                    category_url = "/" + "/".join(category_urlArray[-3:])
                    category_fullUrl = f"{character_achievements_url}{category_url}"

                    # Ajout des tâches pour récupérer les sous-catégories
                    tasks.append(
                        get_character_achievements_subcategories(
                            session, category_fullUrl, category_name
                        )
                    )

            results = await asyncio.gather(*tasks)

            # Fusion des résultats
            for category_name, subcategories in results:
                achievements[category_name] = subcategories

            return achievements

    except ClientResponseError as error:
        raise error

    except Exception as error:
        print(error.__class__)
        print("Error while scraping character's achievements: ", error)
        raise Exception(f"Scraping failed: {str(error)}")


async def get_character_achievements_subcategories(
    session, subCategory_url: str, category_name: str
):
    """
    Fetch subcategories asynchronously
    """
    response = await fetch(session, subCategory_url)
    if response:
        soup = BeautifulSoup(response, "html.parser")

    subCategories = {}

    for entry_category_achievement in soup.select(
        "nav.achievement__category .btn__category__list li"
    ):
        subcategory_elem = entry_category_achievement.select_one("a")
        if subcategory_elem:
            subcategory_name = subcategory_elem.get_text(strip=True)
            subCategories[subcategory_name] = []

    return category_name, subCategories


async def fetch(session, url):
    async with session.get(url) as response:
        if response.status == HTTPStatus.OK:
            return await response.text()
        else:
            response.raise_for_status()
